### ISSUE TITLE
Semantics highlighting gets shifted

### ISSUE THREAD

**Issue by dainiusm07** [dart-lang/sdk#55084]
### Problem
Semantics highlighting gets shifted when I'm entering normal mode in Neovim, I believe it is somehow triggered by the whitespaces that end up being deleted.

UPDATE:
On 3.2.6 everything is working fine. It started appearing only on 3.3.0 and still reproducible on 3.4.0-99.1.beta

https://github.com/dart-lang/sdk/assets/44223680/f3badc4f-6c65-4918-af59-ec46836b9f34

### Dart Info
- Dart 3.3.0 (stable) (Tue Feb 13 10:25:19 2024 +0000) on "macos_arm64"
- on macos / Version 14.2.1 (Build 23C71)

### LSP logs
https://gist.github.com/dainiusm07/e3e43a5499cdda39e9b457caf45aeccc

**Comment by nshahan** [comment]
@natebosch Is there a better repo to triage this issue to?

**Comment by natebosch** [comment]
If the issue is indeed due to a change in behavior between SDK versions 3.2 and 3.3 this is the right repo.

@dainiusm07 - can you provide a log for when highlighting works with an older SDK so we can see the difference between that one and the 3.3 SDK where it is broken?

From the video I'm surprised this would be caused by something happening in the SDK, but seeing a diff in the exact LSP messages from different SDK versions would narrow it down.

**Comment by dainiusm07** [comment]
@natebosch Sure!

### LSP logs of SDK 3.2.6
https://gist.github.com/dainiusm07/9959218a9d70c5db38e7c0b9eaf1d5d0

Updated the original logs, to make sure that I did exactly same actions in both cases.

**Comment by natebosch** [comment]
It does look like the new version of the SDK changed behavior.

In both logs I see:
- a `textDocument/semanticTokens/full` request that receives the same result
- two `textDocument/didChange` messages
- another `textDocument/semanticTokens/full` request
- some `dart/textDocument/publishClosingLabels`, `dart/textDocument/publishOutline`, and a `$/progress`

The two logs diverge in two interesting ways
- The _failing_ log shows some `textDocument/publishDiagnostics` with empty diagnostics. The successful log has no such calls. (Was there a change to always send empty diagnostics instead of omitting the message or something?)
- The _failing_ log has exactly the same data in the `textDocument/semanticTokens/full` response. The successful log has different data (which likely reflects the `didChange` but I didn't check that it was exactly correct).

@bwilkerson - were there recent changes that could have broken the behavior of `didChange` messages, or break prior guarantees about the order that messages get processed?

**Comment by hauserx** [comment]
This change triggered this behavior:

https://dart-review.googlesource.com/c/sdk/+/330100

In particular, building current main with removed below lines fixed the issue:

https://github.com/dart-lang/sdk/blob/51ea22f13f5353c53ab6b5b72e0d94bc7714e20d/pkg/analysis_server/lib/src/analysis_server.dart#L624-L628

CC @DanTup

**Comment by DanTup** [comment]
> The failing log shows some `textDocument/publishDiagnostics` with empty diagnostics. The successful log has no such calls. (Was there a change to always send empty diagnostics instead of omitting the message or something?)

I made an optimisation recently to not send empty diagnostics from the LSP server when we know the client already holds no diagnostics (this avoids sending a `publishDiagnostics` for every single file during startup when there are no diagnostics). This is unrelated to the issue here.

@hauserx thanks for narrowing this down! That change was intended to be a performance improvement (reusing an existing `LineInfo` for the file we already have, rather than building a new one from reading the content) but I wouldn't have expected any behaviour change. I'll try to repro and do some digging.

**Comment by DanTup** [comment]
From your log, it seems like the semantic tokens are being returned based on the first `didChange` and not the second (which seems like a bug because we should always have applied any file buffer changes synchronously before starting any requests that come after it). However, I can't reproduce this in either a unit or integration test doing something similar:

```dart
// Build some content to test with. We'll apply edits adding a character at
    // a time.
    var fullContent = '// ${'A' * 200}';
    newFile(mainFilePath, '');
    await initialize();

    unawaited(openFile(mainFileUri, ''));
    // Loop over the string applying a substring up until the full length.
    // Don't await anything.
    for (var i = 3; i <= fullContent.length; i++) {
      unawaited(replaceFile(i, mainFileUri, fullContent.substring(0, i)));
    }

    // Now the last edit has been sent (but not awaited), fetch the tokens and
    // verify that the length of the only token is the whole content we expect.
    var tokens = await getSemanticTokens(mainFileUri);
    var decoded = decodeSemanticTokens(fullContent, tokens);
    expect(decoded.single.content, fullContent);
```

Are you able to reproduce the issue while recording a server instrumentation log (start the server with `dart language-server --instrumentation-log-file=/path/to/log.txt`) to see if that's any more revealing.

(@hauserx and if you're easily able to reproduce this in a script or something like you did in https://github.com/dart-lang/sdk/issues/55889, that would be awesome too :-))

**Comment by hauserx** [comment]
## Reproduction scenario

Initial dart file - spaces and eols are shown, note the 2 space indent:
![image](https://github.com/dart-lang/sdk/assets/32371993/ca69e1bb-4846-4399-8ab8-1545132fbe36)

After pressing "o" in nvim: adds empty line 2 with same 2 space ident
![image](https://github.com/dart-lang/sdk/assets/32371993/cfda8294-4bdc-44fc-8797-64b0e9db4af7)

After pressing "ESC" it exits edit mode and removes 2 spaces in line 2:
![image](https://github.com/dart-lang/sdk/assets/32371993/ac6e39e2-e9ea-4dea-b724-97cb1cffa613)

**NOTES**: 
  - Letter r in line 3 has wrong highlight
  - Nvim after removing those two spaces sends two separate didChange messages, removing each space separately


## Script
[test_dart_lsp_highlight_shift.py.txt](https://github.com/user-attachments/files/15605192/test_dart_lsp_highlight_shift.py.txt)


Test script sends the same messages as above example using nvim, but sends additional request to send semantic tokens after a second.

Most important parts from log is below, where the first token response has 4th token shifted, but the delayed one is OK:


```
                                                        
                                                1              2               3               4XXXXXXXX      5               6
Received:
--------
{'id': 7, 'jsonrpc': '2.0', 'result': {'data': [0, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0, 2, 1, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0]}}
--------

Received:
--------
{'id': 8, 'jsonrpc': '2.0', 'result': {'data': [0, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0, 2, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0]}}
--------
```

Sometimes the first response is wrong in a different way (looks like timing factors):
```
                                                1              2               3               4XXXXXXXX      5               6
{'id': 7, 'jsonrpc': '2.0', 'result': {'data': [0, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0, 2, 0, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0]}}
```


To run script:
```
./test_dart_lsp_highlight_shift.py /home/hauser/apps/dart/3.3.4/dart-sdk/bin/dart language-server --protocol=lsp
```

<details>
  <summary>Full log</summary>

```
Sending:
----
{"jsonrpc": "2.0", "method": "initialize", "params": {"processId": null, "rootPath": null, "rootUri": "file:///fakepath", "capabilities": {}}, "id": 1}
----

Received:
--------
{'id': 1, 'jsonrpc': '2.0', 'result': {'capabilities': {'callHierarchyProvider': True, 'codeActionProvider': True, 'colorProvider': {'documentSelector': [{'language': 'dart', 'scheme': 'file'}]}, 'completionProvider': {'completionItem': {'labelDetailsSupport': True}, 'resolveProvider': True, 'triggerCharacters': ['.', '=', '(', '$', '"', "'", '{', '/', ':']}, 'definitionProvider': True, 'documentFormattingProvider': True, 'documentHighlightProvider': True, 'documentLinkProvider': {'resolveProvider': False}, 'documentOnTypeFormattingProvider': {'firstTriggerCharacter': '}', 'moreTriggerCharacter': [';']}, 'documentRangeFormattingProvider': True, 'documentSymbolProvider': True, 'executeCommandProvider': {'commands': ['edit.sortMembers', 'edit.organizeImports', 'edit.fixAll', 'edit.sendWorkspaceEdit', 'refactor.perform', 'refactor.validate', 'dart.logAction', 'dart.refactor.convert_all_formal_parameters_to_named', 'dart.refactor.convert_selected_formal_parameters_to_named', 'dart.refactor.move_selected_formal_parameters_left', 'dart.refactor.move_top_level_to_file'], 'workDoneProgress': True}, 'foldingRangeProvider': True, 'hoverProvider': True, 'implementationProvider': True, 'inlayHintProvider': {'resolveProvider': False}, 'referencesProvider': True, 'renameProvider': True, 'selectionRangeProvider': True, 'semanticTokensProvider': {'full': {'delta': False}, 'legend': {'tokenModifiers': ['documentation', 'constructor', 'declaration', 'importPrefix', 'instance', 'static', 'escape', 'annotation', 'control', 'label', 'interpolation', 'void'], 'tokenTypes': ['annotation', 'keyword', 'class', 'comment', 'method', 'variable', 'parameter', 'enum', 'enumMember', 'type', 'source', 'property', 'namespace', 'boolean', 'number', 'string', 'function', 'typeParameter']}, 'range': True}, 'signatureHelpProvider': {'retriggerCharacters': [','], 'triggerCharacters': ['(']}, 'textDocumentSync': {'change': 2, 'openClose': True, 'willSave': False, 'willSaveWaitUntil': False}, 'typeDefinitionProvider': True, 'typeHierarchyProvider': True, 'workspace': {'fileOperations': {'willRename': {'filters': [{'pattern': {'glob': '**/*.dart', 'matches': 'file'}, 'scheme': 'file'}, {'pattern': {'glob': '**/', 'matches': 'folder'}, 'scheme': 'file'}]}}, 'workspaceFolders': {'changeNotifications': True, 'supported': True}}, 'workspaceSymbolProvider': True}, 'serverInfo': {'name': 'Dart SDK LSP Analysis Server', 'version': '3.3.4'}}}
--------

Sending:
----
{"jsonrpc": "2.0", "method": "initialized", "params": {}}
----

Sending:
----
{"jsonrpc": "2.0", "method": "textDocument/semanticTokens/full", "params": {"textDocument": {"uri": "file:///tmp/hello_world.dart"}}, "id": 2}
----

Received:
--------
{'jsonrpc': '2.0', 'method': '$/analyzerStatus', 'params': {'isAnalyzing': True}}
--------

Received:
--------
{'id': 2, 'jsonrpc': '2.0', 'result': {'data': [0, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0, 1, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0]}}
--------

Sending:
----
{"jsonrpc": "2.0", "method": "textDocument/didOpen", "params": {"textDocument": {"uri": "file:///tmp/hello_world.dart", "version": 0, "languageId": "dart", "text": "  var a = 1;\n  var b = 2;\n"}}, "id": 3}
----

Received:
--------
{'id': 3, 'jsonrpc': '2.0', 'result': None}
--------

Sending:
----
{"jsonrpc": "2.0", "method": "textDocument/didChange", "params": {"textDocument": {"uri": "file:///tmp/hello_world.dart", "version": 4}, "contentChanges": [{"range": {"start": {"line": 0, "character": 12}, "end": {"line": 1, "character": 0}}, "rangeLength": 1, "text": "\n  \n"}]}, "id": 4}
----

Received:
--------
{'id': 4, 'jsonrpc': '2.0', 'result': None}
--------

Sending:
----
{"jsonrpc": "2.0", "method": "textDocument/didChange", "params": {"textDocument": {"uri": "file:///tmp/hello_world.dart", "version": 5}, "contentChanges": [{"range": {"start": {"line": 1, "character": 1}, "end": {"line": 1, "character": 2}}, "rangeLength": 1, "text": ""}]}, "id": 5}
----

Received:
--------
{'id': 5, 'jsonrpc': '2.0', 'result': None}
--------

Sending:
----
{"jsonrpc": "2.0", "method": "textDocument/didChange", "params": {"textDocument": {"uri": "file:///tmp/hello_world.dart", "version": 6}, "contentChanges": [{"range": {"start": {"line": 1, "character": 0}, "end": {"line": 1, "character": 1}}, "rangeLength": 1, "text": ""}]}, "id": 6}
----

Sending:
----
{"jsonrpc": "2.0", "method": "textDocument/semanticTokens/full", "params": {"textDocument": {"uri": "file:///tmp/hello_world.dart"}}, "id": 7}
----

Sending:
----
{"jsonrpc": "2.0", "method": "textDocument/semanticTokens/full", "params": {"textDocument": {"uri": "file:///tmp/hello_world.dart"}}, "id": 8}
----

Received:
--------
{'id': 6, 'jsonrpc': '2.0', 'result': None}
--------

Received:
--------
{'jsonrpc': '2.0', 'method': '$/analyzerStatus', 'params': {'isAnalyzing': False}}
--------

Received:
--------
{'id': 7, 'jsonrpc': '2.0', 'result': {'data': [0, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0, 2, 1, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0]}}
--------

Received:
--------
{'id': 8, 'jsonrpc': '2.0', 'result': {'data': [0, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0, 2, 2, 3, 1, 0, 0, 4, 1, 11, 4, 0, 4, 1, 14, 0]}}
```  
</details>

**Comment by DanTup** [comment]
@hauserx thanks! Is there any chance you could run the script with the instrumentation log enabled? ` language-server --protocol=lsp --instrumentation-log-file="C:\Dev\Test Projects\dart_55084\log.txt"`

I'm having trouble running the script - it sends the first messages to `stdin` but it never shows up on the server. I'm on Windows, I don't know if that's a difference. I'll try porting reproducing exactly what's in your script in Dart, but a log captured from the server may include some additional info that might be useful in the meantime.

Thanks!

**Comment by DanTup** [comment]
@hauserx while copying this into the Dart test, I'm a little confused by `wait_for_response=True` in your script. It appears that this would cause a loop waiting for a matching response, but you have it on the `didChange` notifications which are notifications with no ID and not response. It's not clear to me why the script would continue past the first one.

**Comment by DanTup** [comment]
>  I'm a little confused by `wait_for_response=True` in your script.

Oh, I think I see what's happening there... Your client is sending notifications like `didChange` as _requests_ (they have IDs), so the server is generating a response for them. Probably the server should ignore ids on notifications and never send responses for them, but I'm not sure whether it's worth making that change (sending IDs for notifications is out-of-spec so responding in this way is probably at least as useful as just rejecting the request).

In any case, the original log shows your real client isn't doing that so I think it's unrelated to the issue here.

That said, I haven't been able to repro yet. Here's how my server log currently looks:

```
1717670938553:Res:{"id"::0,"jsonrpc"::"2.0","method"::"initialize","params"::{"capabilities"::{"experimental"::{},"textDocument"::{},"window"::{},"workspace"::{}},"processId"::null,"rootPath"::"C::\\fakepath\\","rootUri"::null}}
1717670945239:Req:{"id"::0,"jsonrpc"::"2.0","result"::{"capabilities"::{"callHierarchyProvider"::true,"codeActionProvider"::true,"codeLensProvider"::{},"colorProvider"::{"documentSelector"::[{"language"::"dart","scheme"::"file"}]},"completionProvider"::{"completionItem"::{"labelDetailsSupport"::true},"resolveProvider"::true,"triggerCharacters"::[".","=","(","$","\"","'","{","/","::"]},"definitionProvider"::true,"documentFormattingProvider"::true,"documentHighlightProvider"::true,"documentLinkProvider"::{"resolveProvider"::false},"documentOnTypeFormattingProvider"::{"firstTriggerCharacter"::"}","moreTriggerCharacter"::[";"]},"documentRangeFormattingProvider"::true,"documentSymbolProvider"::true,"executeCommandProvider"::{"commands"::["dart.edit.sortMembers","dart.edit.organizeImports","dart.edit.fixAll","dart.edit.fixAllInWorkspace.preview","dart.edit.fixAllInWorkspace","dart.edit.sendWorkspaceEdit","refactor.perform","refactor.validate","dart.logAction","dart.refactor.convert_all_formal_parameters_to_named","dart.refactor.convert_selected_formal_parameters_to_named","dart.refactor.move_selected_formal_parameters_left","dart.refactor.move_top_level_to_file"],"workDoneProgress"::true},"experimental"::{"textDocument"::{"super"::{},"augmented"::{},"augmentation"::{}}},"foldingRangeProvider"::true,"hoverProvider"::true,"implementationProvider"::true,"inlayHintProvider"::{"resolveProvider"::false},"referencesProvider"::true,"renameProvider"::true,"selectionRangeProvider"::true,"semanticTokensProvider"::{"full"::{"delta"::false},"legend"::{"tokenModifiers"::["documentation","constructor","declaration","importPrefix","instance","static","escape","annotation","control","label","interpolation","void"],"tokenTypes"::["annotation","keyword","class","comment","method","variable","parameter","enum","enumMember","type","source","property","namespace","boolean","number","string","function","typeParameter"]},"range"::true},"signatureHelpProvider"::{"retriggerCharacters"::[","],"triggerCharacters"::["("]},"textDocumentSync"::{"change"::2,"openClose"::true,"willSave"::false,"willSaveWaitUntil"::false},"typeDefinitionProvider"::true,"typeHierarchyProvider"::true,"workspace"::{"fileOperations"::{"willRename"::{"filters"::[{"pattern"::{"glob"::"**/*.dart","matches"::"file"},"scheme"::"file"},{"pattern"::{"glob"::"**/","matches"::"folder"},"scheme"::"file"}]}},"workspaceFolders"::{"changeNotifications"::true,"supported"::true}},"workspaceSymbolProvider"::true},"serverInfo"::{"name"::"Dart SDK LSP Analysis Server","version"::"3.5.0-edge.81f5879ca0bec7d256d7d92038faacc4692cfdb8"}}}
1717670945252:Res:{"jsonrpc"::"2.0","method"::"initialized","params"::{}}
1717670945253:Res:{"id"::1,"jsonrpc"::"2.0","method"::"textDocument/semanticTokens/full","params"::{"textDocument"::{"uri"::"file::///C::/Users/danny/AppData/Local/Temp/analysisServer48c0cdb6/lib/main.dart"}}}
1717670945311:Req:{"jsonrpc"::"2.0","method"::"$/analyzerStatus","params"::{"isAnalyzing"::true}}
1717670945317:Req:{"jsonrpc"::"2.0","method"::"$/analyzerStatus","params"::{"isAnalyzing"::false}}
1717670945318:Req:{"id"::1,"jsonrpc"::"2.0","result"::{"data"::[]}}
1717670945319:Res:{"jsonrpc"::"2.0","method"::"textDocument/didOpen","params"::{"textDocument"::{"languageId"::"dart","text"::"  var a = 1;\n  var b = 2;\n","uri"::"file::///C::/Users/danny/AppData/Local/Temp/analysisServer48c0cdb6/lib/main.dart","version"::0}}}
1717670945319:Res:{"jsonrpc"::"2.0","method"::"textDocument/didChange","params"::{"contentChanges"::[{"range"::{"end"::{"character"::0,"line"::1},"start"::{"character"::12,"line"::0}},"rangeLength"::1,"text"::"\n  \n"}],"textDocument"::{"uri"::"file::///C::/Users/danny/AppData/Local/Temp/analysisServer48c0cdb6/lib/main.dart","version"::4}}}
1717670945320:Res:{"jsonrpc"::"2.0","method"::"textDocument/didChange","params"::{"contentChanges"::[{"range"::{"end"::{"character"::2,"line"::1},"start"::{"character"::1,"line"::1}},"rangeLength"::1,"text"::""}],"textDocument"::{"uri"::"file::///C::/Users/danny/AppData/Local/Temp/analysisServer48c0cdb6/lib/main.dart","version"::5}}}
1717670945320:Res:{"jsonrpc"::"2.0","method"::"textDocument/didChange","params"::{"contentChanges"::[{"range"::{"end"::{"character"::1,"line"::1},"start"::{"character"::0,"line"::1}},"rangeLength"::1,"text"::""}],"textDocument"::{"uri"::"file::///C::/Users/danny/AppData/Local/Temp/analysisServer48c0cdb6/lib/main.dart","version"::6}}}
1717670945320:Res:{"id"::2,"jsonrpc"::"2.0","method"::"textDocument/semanticTokens/full","params"::{"textDocument"::{"uri"::"file::///C::/Users/danny/AppData/Local/Temp/analysisServer48c0cdb6/lib/main.dart"}}}
1717670945331:Req:{"jsonrpc"::"2.0","method"::"$/analyzerStatus","params"::{"isAnalyzing"::true}}
1717670945605:Req:{"id"::2,"jsonrpc"::"2.0","result"::{"data"::[0,2,3,1,0,0,4,1,11,4,0,4,1,14,0,2,2,3,1,0,0,4,1,11,4,0,4,1,14,0]}}
1717670945612:Res:{"id"::3,"jsonrpc"::"2.0","method"::"textDocument/semanticTokens/full","params"::{"textDocument"::{"uri"::"file::///C::/Users/danny/AppData/Local/Temp/analysisServer48c0cdb6/lib/main.dart"}}}
1717670945629:Req:{"id"::3,"jsonrpc"::"2.0","result"::{"data"::[0,2,3,1,0,0,4,1,11,4,0,4,1,14,0,2,2,3,1,0,0,4,1,11,4,0,4,1,14,0]}}
```

Both of the results at the end look correct and the same. However, I notice in your code you are waiting for the response inside your loop:

```python
for i in range(2):
    send_message(
        server,
        "textDocument/semanticTokens/full",
        {"textDocument": {
            "uri": "file:///C:/Dev/Test Projects/dart_55084/project/hello_world.dart",
        }},
        wait_for_response=False)
    time.sleep(1)
```

However your latest log shows two semantic tokens requests sent consecutively, before the response to the second arrives. Was this log from the same script or are these out of sync?

**Comment by hauserx** [comment]
Changed script and tested it runs under windows (reading some bytes one after another, but it does it's job).

I am able to reproduce the issue on windows, but the issue happens now from time to time instead of always:

[test_dart_lsp_highlight_shift.py.txt](https://github.com/user-attachments/files/15614240/test_dart_lsp_highlight_shift.py.txt)



> I'm a little confused by wait_for_response=True in your script. It appears that this would cause a loop waiting for a matching response, but you have it on the didChange notifications which are notifications with no ID and not response.

Oh, you right here, but the server still returns response for those requests, as you can see in the full log above, e.g.:

> {'id': 4, 'jsonrpc': '2.0', 'result': None}

Just reading jsonrpc spec server is required to return a response if id is in request, but I should not include those in the repro - it makes it more deterministic though.

**Comment by hauserx** [comment]
> However your latest log shows two semantic tokens requests sent consecutively, before the response to the second arrives. Was this log from the same script or are these out of sync?

No, just send_message do not wait for incoming messages and does not print those, so the responses are printed later in while True: read_messages.

**Comment by DanTup** [comment]
> No, just send_message do not wait for incoming messages and does not print those, so the responses are printed later in while True: read_messages.

Oops, I apparently misread `wait_for_response=False` as `True`!

The script runs now, but it seems to always produce the right output for me:

![image](https://github.com/dart-lang/sdk/assets/1078012/1780c2a7-8ac1-4bdb-b27b-fbeeef7ee537)

Are you able to run with the instrumentation log (`--protocol=lsp --instrumentation-log-file="C:\Dev\Test Projects\dart_55084\log.txt"`) and provide that log when this occurs, in case it has anything that might help me better understand the exact timing that's triggering it?

**Comment by hauserx** [comment]
Here is the log:
```
1717673556944:Ver:1717670435143993511749:unknown.client.lsp:unknown:1.37.0:3.4.0
1717673556974:Req:{"jsonrpc":: "2.0", "method":: "initialize", "params":: {"processId":: null, "rootPath":: null, "rootUri":: "file::///c::/ub/dart/mytest", "capabilities":: {}}, "id":: 1}
1717673557003:Res:{"id"::1,"jsonrpc"::"2.0","result"::{"capabilities"::{"callHierarchyProvider"::true,"codeActionProvider"::true,"codeLensProvider"::{},"colorProvider"::{"documentSelector"::[{"language"::"dart","scheme"::"file"}]},"completionProvider"::{"completionItem"::{"labelDetailsSupport"::true},"resolveProvider"::true,"triggerCharacters"::[".","=","(","$","\"","'","{","/","::"]},"definitionProvider"::true,"documentFormattingProvider"::true,"documentHighlightProvider"::true,"documentLinkProvider"::{"resolveProvider"::false},"documentOnTypeFormattingProvider"::{"firstTriggerCharacter"::"}","moreTriggerCharacter"::[";"]},"documentRangeFormattingProvider"::true,"documentSymbolProvider"::true,"executeCommandProvider"::{"commands"::["dart.edit.sortMembers","dart.edit.organizeImports","dart.edit.fixAll","dart.edit.fixAllInWorkspace.preview","dart.edit.fixAllInWorkspace","dart.edit.sendWorkspaceEdit","refactor.perform","refactor.validate","dart.logAction","dart.refactor.convert_all_formal_parameters_to_named","dart.refactor.convert_selected_formal_parameters_to_named","dart.refactor.move_selected_formal_parameters_left","dart.refactor.move_top_level_to_file"],"workDoneProgress"::true},"experimental"::{"textDocument"::{"super"::{},"augmented"::{},"augmentation"::{}}},"foldingRangeProvider"::true,"hoverProvider"::true,"implementationProvider"::true,"inlayHintProvider"::{"resolveProvider"::false},"referencesProvider"::true,"renameProvider"::true,"selectionRangeProvider"::true,"semanticTokensProvider"::{"full"::{"delta"::false},"legend"::{"tokenModifiers"::["documentation","constructor","declaration","importPrefix","instance","static","escape","annotation","control","label","interpolation","void"],"tokenTypes"::["annotation","keyword","class","comment","method","variable","parameter","enum","enumMember","type","source","property","namespace","boolean","number","string","function","typeParameter"]},"range"::true},"signatureHelpProvider"::{"retriggerCharacters"::[","],"triggerCharacters"::["("]},"textDocumentSync"::{"change"::2,"openClose"::true,"willSave"::false,"willSaveWaitUntil"::false},"typeDefinitionProvider"::true,"typeHierarchyProvider"::true,"workspace"::{"fileOperations"::{"willRename"::{"filters"::[{"pattern"::{"glob"::"**/*.dart","matches"::"file"},"scheme"::"file"},{"pattern"::{"glob"::"**/","matches"::"folder"},"scheme"::"file"}]}},"workspaceFolders"::{"changeNotifications"::true,"supported"::true}},"workspaceSymbolProvider"::true},"serverInfo"::{"name"::"Dart SDK LSP Analysis Server","version"::"3.4.3"}}}
1717673557005:Req:{"jsonrpc":: "2.0", "method":: "initialized", "params":: {}}
1717673557033:Res:{"jsonrpc"::"2.0","method"::"$/analyzerStatus","params"::{"isAnalyzing"::true}}
1717673557101:Res:{"jsonrpc"::"2.0","method"::"$/analyzerStatus","params"::{"isAnalyzing"::false}}
1717673558006:Req:{"jsonrpc":: "2.0", "method":: "textDocument/semanticTokens/full", "params":: {"textDocument":: {"uri":: "file::///c::/ub/dart/mytest/test.dart"}}, "id":: 2}
1717673558013:Res:{"jsonrpc"::"2.0","method"::"$/analyzerStatus","params"::{"isAnalyzing"::true}}
1717673558076:Res:{"id"::2,"jsonrpc"::"2.0","result"::{"data"::[0,2,3,1,0,0,4,1,11,4,0,4,1,14,0,1,2,3,1,0,0,4,1,11,4,0,4,1,14,0]}}
1717673558076:Req:{"jsonrpc":: "2.0", "method":: "textDocument/didOpen", "params":: {"textDocument":: {"uri":: "file::///c::/ub/dart/mytest/test.dart", "version":: 0, "languageId":: "dart", "text":: "  var a = 1;\n  var b = 2;\n"}}, "id":: 3}
1717673558084:Res:{"id"::3,"jsonrpc"::"2.0","result"::null}
1717673558091:Req:{"jsonrpc":: "2.0", "method":: "textDocument/didChange", "params":: {"textDocument":: {"uri":: "file::///c::/ub/dart/mytest/test.dart", "version":: 4}, "contentChanges":: [{"range":: {"start":: {"line":: 0, "character":: 12}, "end":: {"line":: 1, "character":: 0}}, "rangeLength":: 1, "text":: "\n  \n"}]}, "id":: 4}
1717673558093:Res:{"id"::4,"jsonrpc"::"2.0","result"::null}
1717673558097:Req:{"jsonrpc":: "2.0", "method":: "textDocument/didChange", "params":: {"textDocument":: {"uri":: "file::///c::/ub/dart/mytest/test.dart", "version":: 5}, "contentChanges":: [{"range":: {"start":: {"line":: 1, "character":: 1}, "end":: {"line":: 1, "character":: 2}}, "rangeLength":: 1, "text":: ""}]}, "id":: 5}
1717673558097:Res:{"id"::5,"jsonrpc"::"2.0","result"::null}
1717673558103:Req:{"jsonrpc":: "2.0", "method":: "textDocument/didChange", "params":: {"textDocument":: {"uri":: "file::///c::/ub/dart/mytest/test.dart", "version":: 6}, "contentChanges":: [{"range":: {"start":: {"line":: 1, "character":: 0}, "end":: {"line":: 1, "character":: 1}}, "rangeLength":: 1, "text":: ""}]}, "id":: 6}
1717673558103:Res:{"id"::6,"jsonrpc"::"2.0","result"::null}
1717673558103:Req:{"jsonrpc":: "2.0", "method":: "textDocument/semanticTokens/full", "params":: {"textDocument":: {"uri":: "file::///c::/ub/dart/mytest/test.dart"}}, "id":: 7}
1717673558117:Res:{"id"::7,"jsonrpc"::"2.0","result"::{"data"::[0,2,3,1,0,0,4,1,11,4,0,4,1,14,0,2,1,3,1,0,0,4,1,11,4,0,4,1,14,0]}}
1717673558229:Res:{"jsonrpc"::"2.0","method"::"$/analyzerStatus","params"::{"isAnalyzing"::false}}
1717673559103:Req:{"jsonrpc":: "2.0", "method":: "textDocument/semanticTokens/full", "params":: {"textDocument":: {"uri":: "file::///c::/ub/dart/mytest/test.dart"}}, "id":: 8}
1717673559103:Res:{"id"::8,"jsonrpc"::"2.0","result"::{"data"::[0,2,3,1,0,0,4,1,11,4,0,4,1,14,0,2,2,3,1,0,0,4,1,11,4,0,4,1,14,0]}}
```

**Comment by DanTup** [comment]
Thanks - I modified your script to run in a loop and stop when the issue occurred and managed to trigger it after around 8 iterations. This is good enough for me to debug some. Thanks!

**Comment by DanTup** [comment]
@bwilkerson @scheglov I _think_ I understand what's happening here now, but I'm not sure whether it's a bug or the assumption made for https://dart-review.googlesource.com/c/sdk/+/330100 was incorrect.

The assumption made was that calling `getAnalysisDriver(path)?.getFileSync(path)` would always return the current/latest version of a file the server knows about (and this would a faster way to get a `LineInfo` than to ask `resourceProvider` for the content and then build a new `LineInfo`).

However, if an overlay has just been updated, it seems that this can return the previous content until the driver has done its next iteration of work (setting the overlay calls `drive.changeFile()` synchronously but that doesn't remove the old file state, it just marks that this file has been changed and schedules future work - which may or may not occur prior to the next incoming client request starting).

So when this bug occurs, I think the order of operations are:

1. Client sends updateOverlay
2. Client sends request for semantic tokens
3. Server processes updateOverlay, which calls driver.changeFile(), which marks that this file needs updating
4. Server processes request for semantic tokens which reads file state
5. Server updates file state

I think the order of 4+5 is undefined because both rely on Dart's scheduling (or things in the incoming stream and the work triggered by Drive.notify()`), and `Driver._run` will call `pumpEventQueue()` a few times at the start of its loop, which allows work like (4) to start being processed.

My feeling is that my original assumption was incorrect, and we should not assume `getAnalysisDriver(path)?.getFileSync(path)` necessarily includes the latest version of the content sent by the client.

That said, I think we have another problem here... For the server-provided tokens we are getting a resolved unit and already have a consistent `LineInfo` we should use that would eliminate this code entirely. The reason we try to get a `LineInfo` outside of that is because we may get results from plugins (and we might not even have a resolved unit from server, if the file type was non-Dart). This (re-)highlights that if the server and plugin are operating on different versions of a file then merging the results from each is not safe.

I wonder if any plugins are actually trying to provide any highlights for Dart files. If not, we could perhaps say that only the server produces them for Dart, and we delegate to plugins for others (it doesn't address the fact that we're still using potentially-invalid `LineInfo`s for plugins, but it at least eliminates the issue for Dart).

**Comment by bwilkerson** [comment]
@srawlins For the discussion of plugins.

> My feeling is that my original assumption was incorrect, and we should not assume `getAnalysisDriver(path)?.getFileSync(path)` necessarily includes the latest version of the content sent by the client.

I believe that's true.

Also, that code path bypasses the check in `AnalysisSession` that ensures that the results returned are consistent with previously returned results.

> For the server-provided tokens we are getting a resolved unit and already have a consistent `LineInfo` we should use that would eliminate this code entirely.

Yes, we should make sure that any code that needs a `LineInfo` can access the one that's already computed for the file.

That might involve creating a protocol-agnostic representation for the results in some cases, but that something I would like to see us pushing toward anyway.

> The reason we try to get a `LineInfo` outside of that is because we may get results from plugins ...
> ...
> I wonder if any plugins are actually trying to provide any highlights for Dart files. If not, we could perhaps say that only the server produces them for Dart, and we delegate to plugins for others ...

I suspect that there aren't any plugins that are trying to send highlights for Dart files, but I have no way of knowing for sure.

That said, I believe the plan for the real plugin support is that we will only be supporting diagnostics (mostly for lints) and fixes (though I suspect we might want to open that up to allow assists as well).

I think it would be valuable for us to disable any support in the current plugin prototype that we don't intend to include in the supported version so that we can tell from the code which features do and do not need to be concerned with data from plugins. (Doing so might involve a breaking change notice being sent out so that we can be aware if any plugins might be impacted by the change.)

**Comment by DanTup** [comment]
> I believe that's true.
> 
> Also, that code path bypasses the check in AnalysisSession that ensures that the results returned are consistent with previously returned results.

I don't _think_ checking the session helps here, because (if I understand correctly) the session still hasn't been invalidated either (we updated the overlay and scheduled work, but nothing more?). But if it had, it would probably result in us throwing/rejecting here, but actually we should be returning correct results.

So I think we should revert https://dart-review.googlesource.com/c/sdk/+/330100 and I now have a test for semantic tokens that fails on current code and passes with that revert.

But, I will also change how semantic tokens works a bit so that we're directly using the `LineInfo` from the resolved unit in the case of a Dart file. This doesn't address that it might be inconsistent for plugin-provided results, that's already the case (but today it's the case both server + plugins, whereas being only the case for plugins is an improvement).


> I think it would be valuable for us to disable any support in the current plugin prototype

I don't know if "current plugin prototype" means the currently-shipped prototype or the updated prototype, but I think I agree either way - the more code using `server.getLineInfo()` (which is the source of this bug) that can be removed, the less scope for bugs like this (and most of the use of that method is because of having to handle plugin results). I don't know if it's well-defined enough to send out a notice, but should we open an issue for this (to send a notice and ultimately remove that support?).

**Comment by bwilkerson** [comment]
> I don't know if "current plugin prototype" means the currently-shipped prototype or the updated prototype ...

I meant the currently-shipped prototype. For the updated "prototype" we're effectively starting from scratch and only adding functionality that we intend to support (assuming that the work is successful and it ships), so there's nothing to remove.

**Comment by scheglov** [comment]
If you want to make sure that latest `changeFile()` notifications were processed, await `applyPendingFileChanges()`.

**Comment by DanTup** [comment]
Ah, that sounds better - I'll update my CL to do that. Thanks!

**Comment by DanTup** [comment]
Oh, I missed that `applyPendingFileChanges` is async and this `getLineInfo()` is not currently. That's a slightly bigger change than I'd like to try and make for this fix (I'd have to review all places using `server.getLineInfo()` to ensure they can be async and an `await` at that position doesn't change any assumptions). I'll leave as-is for now, and if we can't get rid of `server.getLineInfo()` soon (most likely as part of plugin changes), I can look at that.

**Comment by DanTup** [comment]
@dainiusm07 @hauserx the fix for this is in the latest Flutter `master` and the latest [Dart dev release](https://dart.dev/get-dart/archive#dev-channel) if you want to confirm it's working as expected in your client now. If you still see issues, please let me know (ideally with an instrumentation log from the server as detailed above).

Thanks!

**Comment by DanTup** [comment]
Great, thanks for confirming!
